src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+
+use super::batch::{Batch, BatchSettings};
+use super::buffer::partition::Partition;
+use crate::buffers::Acker;
+use futures01::{
+    future::Either,
+    stream::FuturesUnordered,
+    sync::oneshot::{self, Receiver},
+    try_ready, Async, AsyncSink, Future, Poll, Sink, StartSend, Stream,
+};
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt,
+    hash::Hash,
+    marker::PhantomData,
+    time::Instant,
+};
+use tokio01::{
+    executor::{DefaultExecutor, Executor},
+    timer::Delay,
+};
+use tower::Service;
+use tracing_futures::Instrument;
+
+// === StreamSink ===
+
+const STREAM_SINK_MAX: usize = 10_000;
+
+/// Simple stream based sink adapter.
+///
+/// This will wrap any inner sink acking all events
+/// as soon as poll_complete returns ready. `start_send`
+/// will also attempt to fully flush if the amount of
+/// in flight acks is larger than `STREAM_SINK_MAX`.
+#[derive(Debug)]
+pub struct StreamSink<T> {
+    inner: T,
+    acker: Acker,
+    pending: usize,
+}
+
+impl<T> StreamSink<T> {
+    pub fn new(inner: T, acker: Acker) -> Self {
+        Self {
+            inner,
+            acker,
+            pending: 0,
+        }
+    }
+}
+
+impl<T: Sink> Sink for StreamSink<T> {
+    type SinkItem = T::SinkItem;
+    type SinkError = T::SinkError;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        trace!("sending item.");
+        match self.inner.start_send(item)? {
+            AsyncSink::Ready => {
+                self.pending += 1;
+                trace!(message = "submit successful.", pending_acks = self.pending);
+
+                if self.pending >= STREAM_SINK_MAX {
+                    self.poll_complete()?;
+                }
+
+                Ok(AsyncSink::Ready)
+            }
+
+            AsyncSink::NotReady(item) => {
+                trace!("Inner sink applying back pressure.");
+                Ok(AsyncSink::NotReady(item))
+            }
+        }
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        try_ready!(self.inner.poll_complete());
+
+        trace!(message = "Acking events.", acking_num = self.pending);
+        self.acker.ack(self.pending);
+        self.pending = 0;
+
+        Ok(().into())
+    }
+}
+
+// === BatchSink ===
+
+/// A `Sink` interface that wraps a `Service` and a
+/// `Batch`.
+///
+/// Provided a batching scheme, a service and batch settings
+/// this type will handle buffering events via the batching scheme
+/// and dispatching requests via the service based on either the size
+/// of the batch or a batch linger timeout.
+///
+/// # Acking
+///
+/// Service based acking will only ack events when all prior request
+/// batches have been acked. This means if sequential requests r1, r2,
+/// and r3 are dispatched and r2 and r3 complete, all events contained
+/// in all requests will not be acked until r1 has completed.
+pub struct BatchSink<S, B, Request, E = DefaultExecutor> {
```

Comment:
I would be totally okay with these being separate files.

Reply:
I am happy to do that, the big reason I decided to keep it one file is because this is just utility code. @lukesteensen any preference?

Reply:
It's just a lot of code!

Reply:
I agree it's a lot of code, but I don't feel too strongly either way.